crates/phactory/src/system/mod.rs

```diff
@@ -632,24 +694,65 @@ impl<Platform: pal::Platform> System<Platform> {
             .process_event(block, event, &mut WorkerSMDelegate(&self.egress), true);
     }
 
-    fn set_master_key(&mut self, master_key: sr25519::Pair, need_restart: bool) {
+    /// Update sealing keys if the received history is longer than existing one. This will not affect the master key history
+    /// in `self.gatekeeper`.
+    ///
+    /// Only restart if the flag is set and the master key is changed
+    fn set_master_key_history(
+        &mut self,
+        master_key_history: Vec<Sr25519SecretKey>,
+        need_restart: bool,
+    ) {
+        if master_key_history.len() <= self.master_key_history.len() {
+            return;
+        }
+
+        master_key::seal(
+            self.sealing_path.clone(),
+            &master_key_history,
+            &self.identity_key,
+            &self.platform,
+        );
+        self.master_key_history = master_key_history;
+
+        let first_key = self.master_key_history.first().expect("check; qed");
         if self.master_key.is_none() {
+            self.master_key = Some(sr25519::Pair::restore_from_secret_key(&first_key));
+
+            if need_restart {
+                crate::maybe_remove_checkpoints(&self.sealing_path);
+                panic!(
+                    "Received master key, please restart pRuntime and pherry to sync as Gatekeeper"
+                );
+            }
+        }
+    }
+
+    // This will change `self.gatekeeper`'s master key history and sequence id since the gatekeeper will report the new
+    // master pubkey on-chain.
+    fn handle_new_master_key(&mut self, new_master_key: sr25519::Pair, rotation_id: u64) {
+        let new_secret = new_master_key.dump_secret_key();
+        if !self.master_key_history.contains(&new_secret) {
```

Comment:
The `contains` is a little bit confusing, maybe better give each key a sequence number?

Reply:
Fixed.